### ISSUE TITLE
Does `st_wrap_dateline()` work with S2?

### ISSUE THREAD

**Issue by kadyb** [r-spatial/sf#2250]
I wanted to reproduce [this example](https://github.com/rspatial/terra/issues/1330) in `{sf}` using `{s2}` and I have two issues:

1. It looks like `st_wrap_dateline()` doesn't work for `{s2}`?
2. Shouldn't the convex hull look different than the one generated by GEOS (e.g. lines between points more oval)?

```r
library("sf")

url = "https://github.com/rspatial/terra/files/13204234/opportunistic-occ2.csv"
tmp = tempfile(fileext = ".csv")
download.file(url, tmp)

locs = read.csv(tmp, row.names = 1)
locs = st_as_sf(locs, coords = 1:2, crs = "+proj=longlat")
mcp = st_convex_hull(st_combine(locs))

d = units::set_units(1000, km)
b = st_buffer(mcp, dist = d)
b = st_wrap_dateline(b, options = c("WRAPDATELINE=YES", "DATELINEOFFSET=180"))

maps::map("world", col = "grey85", fill = TRUE, border = "grey85")
plot(locs, pch = 20, col = "red", add = TRUE)
plot(mcp, add = TRUE)
plot(b, border = "blue", lwd = 2, add = TRUE)
```

![Rplot](https://github.com/r-spatial/sf/assets/35004826/c0a0128b-9826-4fe0-a1d4-9ccaf8907a35)

**Comment by kadyb** [comment]
1. S2 must be turned off to work: https://github.com/r-spatial/sf/issues/1999
2. `st_convex_hull()` only [uses](https://github.com/r-spatial/sf/issues/1790#issuecomment-916820577) GEOS, even though a spherical version [is implemented](https://r-spatial.github.io/s2/reference/s2_boundary.html) in `{s2}`?

**Comment by edzer** [comment]
`st_wrap_dateline` works if the buffer is computed by GEOS:
```r
sf_use_s2(FALSE)
d = 10 # now degrees
b = st_buffer(mcp, dist = d)
bb = st_wrap_dateline(b, options = c("WRAPDATELINE=YES"))                             

maps::map("world", col = "grey85", fill = TRUE, border = "grey85")                    
plot(locs, pch = 20, col = "red", add = TRUE)                                         
plot(mcp, add = TRUE)
plot(bb, border = "blue", lwd = 2, add = TRUE) 
```
I have no clue why. Will look into convex hull on s2.

**Comment by rsbivand** [comment]
I think `s2` understands coordinates as spherical intrinsically, but does it see longitude `-181` as `+179` automatically?

**Comment by edzer** [comment]
```r
> st_as_s2(locs) |> s2_convex_hull_agg()                                                                                                                               
<geodesic s2_geography[1] with CRS=OGC:CRS84>                                                                                                                          
[1] POLYGON ((0 -90, 0 -90))  
```
i.e., the full polygon (the entire sphere). As opposed to e.g.
```r
> st_sfc(st_point(0:1), st_point(1:0), st_point(c(1,1)), crs = st_crs("OGC:CRS84")) |> st_as_s2() |> s2_convex_hull_agg()                                              
<geodesic s2_geography[1] with CRS=OGC:CRS84>                                                                                                                          
[1] POLYGON ((1 0, 1 1, 0 1, 1 0))  
```

**Comment by edzer** [comment]
> I think s2 understands coordinates as spherical intrinsically, but does it see longitude -181 as +179 automatically?

Yes.

**Comment by edzer** [comment]
> I wonder one more thing -- was the buffer generated correctly on the sphere at all? It think the distance from central Africa to Poland is more than 1000 km.

Yes; the issue is that a straight line on plate carree (red) is not a geodesic / great circle (blue, obtained with `st_segmentise`):

![x](https://github.com/r-spatial/sf/assets/520851/416fa943-7f4c-4f07-bf93-894e40243604)

Also note that this "convex hull" is a convex hull in plate carree, but apparently not on the sphere.

**Comment by edzer** [comment]
> I have no clue why.

I now do. See [this comment](https://github.com/OSGeo/gdal/blob/master/ogr/ogrgeometryfactory.cpp#L2971-L2972): gdal's wrapdateline skips a geometry if it has a longitude exactly equal to -180 or 180, which `b` does:
```r
> st_bbox(b)                                                                                                                                                           
      xmin       ymin       xmax       ymax                                                                                                                            
-180.00000  -62.07510  179.39853   68.63198       
```
If we hack this, e.g. by
```r
i = which(b[[1]][[1]][,1] == -180)                                                                                                                                   
b[[1]][[1]][i,1] = -179.999 # note that the bbox attribute is now not updated!
```
we get
```r
> st_wrap_dateline(b, options = c("WRAPDATELINE=YES"))                                                                                                                 
Geometry set for 1 feature                                                                                                                                             
Geometry type: MULTIPOLYGON                                                                                                                                            
Dimension:     XY                                                                                                                                                      
Bounding box:  xmin: -180 ymin: -62.0751 xmax: 180 ymax: 68.63198                                                                                                      
Geodetic CRS:  WGS 84                                                                                                                                                  
MULTIPOLYGON (((178.7879 -33.07304, 178.7879 -3...    
```
in plot:
![x](https://github.com/r-spatial/sf/assets/520851/ea377299-5608-4951-9f2b-33aadf8c617e)

**Comment by edzer** [comment]
Solved in GDAL 3.8.0beta1

**Comment by edzer** [comment]
> st_convex_hull() only https://github.com/r-spatial/sf/issues/1790#issuecomment-916820577 GEOS, even though a spherical version [is implemented](https://r-spatial.github.io/s2/reference/s2_boundary.html) in {s2}?

@kadyb this is now fixed, too.
